File: spec/java/src/io/kaitai/struct/specwrite/TestCastToTop.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ByteBufferKaitaiStream;
import io.kaitai.struct.KaitaiStream;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.CastToTop;
import org.testng.annotations.Test;

public class TestCastToTop extends CommonSpec {
    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return CastToTop.class;
    }

    @Override
    protected String getSrcFilename() {
        return "fixed_struct.bin";
    }

    @Test
    public void testInstDisabled() throws Exception {
        CastToTop r = new CastToTop();
        r.setCode(0x77);

        CastToTop s = new CastToTop(null, r, r._root());
        s.setCode(0xa8);
        // `cast_to_top.ksy` is an infinitely recursive format, so it's necessary to break
        // the infinite recursion by disabling the `header` instance here.
        s.setHeader_Enabled(false);
        s._check();

        r.setHeader(s);
        r._check();

        byte[] output = new byte[2];
        try (KaitaiStream io = new ByteBufferKaitaiStream(output)) {
            r._write(io);
        }
        assertByteArrayEquals(output, new byte[] { 0x77, (byte) 0xa8 });
    }
}

File: spec/python/specwrite/test_cast_to_top.py
import io
import unittest
from kaitaistruct import KaitaiStream
from specwrite.common_spec import CommonSpec

from testwrite.cast_to_top import CastToTop

class TestCastToTop(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestCastToTop, self).__init__(*args, **kwargs)
        self.struct_class = CastToTop
        self.src_filename = 'src/fixed_struct.bin'

    def test_inst_disabled(self):
        r = CastToTop()
        r.code = 0x77

        s = CastToTop(None, r, r._root)
        s.code = 0xa8
        # `cast_to_top.ksy` is an infinitely recursive format, so it's necessary to break
        # the infinite recursion by disabling the `header` instance here.
        s.header__enabled = False
        s._check()

        r.header = s
        r._check()

        ks_io = KaitaiStream(io.BytesIO(bytearray(2)))
        r._write(ks_io)

        actual = ks_io.to_byte_array()
        self.assert_byte_array_equal(actual, b"\x77\xa8")
